### PR TITLE
Lebab transformed

### PR DESCRIPTION
**Pull request by MurakamiShinyu**

### CODE REVIEW

File: src/vivliostyle/urls.js [hunk 1]
```diff
@@ -26,12 +26,4 @@ goog.provide("vivliostyle.urls");
...
-    return attributeValue.replace(/[uU][rR][lL]\(\s*\"((\\([^0-9a-fA-F]+|[0-9a-fA-F]+\s*)|[^\"\r\n])+)\"/gm, function(match, m1) {
-        return 'url("' + documentURLTransformer.transformURL(m1, baseUrl)+ '"';
-    }).replace(/[uU][rR][lL]\(\s*\'((\\([^0-9a-fA-F]+|[0-9a-fA-F]+\s*)|[^\'\r\n])+)\'/gm, function(match, m1) {
-        return "url('" + documentURLTransformer.transformURL(m1, baseUrl)+ "'";
-    }).replace(/[uU][rR][lL]\(\s*((\\([^0-9a-fA-F]+|[0-9a-fA-F]+\s*)|[^\"\'\r\n\)\s])+)/gm, function(match, m1) {
-        return "url(" + documentURLTransformer.transformURL(m1, baseUrl);
-    });
-};
+vivliostyle.urls.transformURIs = (attributeValue, baseUrl, documentURLTransformer) => attributeValue.replace(/[uU][rR][lL]\(\s*\"((\\([^0-9a-fA-F]+|[0-9a-fA-F]+\s*)|[^\"\r\n])+)\"/gm, (match, m1) => `url("${documentURLTransformer.transformURL(m1, baseUrl)}"`).replace(/[uU][rR][lL]\(\s*\'((\\([^0-9a-fA-F]+|[0-9a-fA-F]+\s*)|[^\'\r\n])+)\'/gm, (match, m1) => `url('${documentURLTransformer.transformURL(m1, baseUrl)}'`).replace(/[uU][rR][lL]\(\s*((\\([^0-9a-fA-F]+|[0-9a-fA-F]+\s*)|[^\"\'\r\n\)\s])+)/gm, (match, m1) => `url(${documentURLTransformer.transformURL(m1, baseUrl)}`);
```
Comment (spring-raining) [inline src/vivliostyle/urls.js]: It's no suitable code for a human!

File: src/vivliostyle/util.js [hunk 2]
```diff
@@ -19,7 +19,7 @@
 /* eslint no-extend-native: "off" */
 goog.provide("vivliostyle.util");
 
-(function() {
+((() => {
```
Comment (spring-raining) [inline src/vivliostyle/util.js]: It's seemed to have too many parentheses.

**Review by spring-raining — COMMENTED** [review]
The Lebab transformation seems to be very good results. However, some type annotations using JSDoc are dropped therefore the typecheck of Closure Compiler raises many warnings.

**Comment by spring-raining** [comment]
### Details of Closure Compiler typecheck
<details>

```
$ java -jar node_modules/google-closure-compiler/compiler.jar --checks_only --compilation_level ADVANCED_OPTIMIZATIONS --language_in ECMASCRIPT6 --warning_level VERBOSE --externs src/externs.js --process_common_js_modules --hide_warnings_for node_modules/fast-diff/diff.js src/closure/goog/base.js src/closure/goog/*/*.js src/adapt/*.js src/vivliostyle/*.js plugins/*/src/*.js node_modules/fast-diff/diff.js
src/adapt/base.js:684: WARNING - actual parameter 1 of $jscomp.makeIterator does not match formal parameter
found   : (Array<T>|null)
required: (Arguments<?>|Array<?>|Iterable<?>|Iterator<?>|string)
    for (const v of arr) {
                    ^

src/adapt/base.js:721: WARNING - actual parameter 1 of $jscomp.makeIterator does not match formal parameter
found   : (Array<T>|null)
required: (Arguments<?>|Array<?>|Iterable<?>|Iterator<?>|string)
    for (const v of arr) {
                    ^

src/adapt/csscasc.js:359: WARNING - Misplaced type annotation. Type annotations are not allowed here. Are you missing parentheses?
adapt.csscasc.getProp = (style, name) => /** @type {adapt.csscasc.CascadeValue} */ style[name];
                                                                                   ^

src/adapt/csscasc.js:381: WARNING - Misplaced type annotation. Type annotations are not allowed here. Are you missing parentheses?
adapt.csscasc.getStyleMap = (style, name) => /** @type {adapt.csscasc.ElementStyleMap} */ style[name];
                                                                                          ^

src/adapt/csscasc.js:415: WARNING - Misplaced type annotation. Type annotations are not allowed here. Are you missing parentheses?
adapt.csscasc.getSpecial = (style, name) => /** @type {Array.<adapt.csscasc.CascadeValue>} */ style[name];
                                                                                              ^

src/adapt/csscasc.js:3882: WARNING - actual parameter 1 of $jscomp.makeIterator does not match formal parameter
found   : (Array<string>|null)
required: (Arguments<string>|Array<string>|Iterable<string>|Iterator<string>|string)
        for (const regionId of regionIds) {
                               ^

src/adapt/cssvalid.js:208: WARNING - actual parameter 1 of $jscomp.arrayFromIterable does not match formal parameter
found   : (Array<number>|null)
required: (Arguments<number>|Array<number>|Iterable<number>|string)
    group.match.push(...this.match);
                        ^

src/adapt/cssvalid.js:209: WARNING - actual parameter 1 of $jscomp.arrayFromIterable does not match formal parameter
found   : (Array<number>|null)
required: (Arguments<number>|Array<number>|Iterable<number>|string)
    group.nomatch.push(...this.nomatch);
                          ^

src/adapt/cssvalid.js:210: WARNING - actual parameter 1 of $jscomp.arrayFromIterable does not match formal parameter
found   : (Array<number>|null)
required: (Arguments<number>|Array<number>|Iterable<number>|string)
    group.error.push(...this.error);
                        ^

src/adapt/cssvalid.js:404: WARNING - actual parameter 1 of $jscomp.makeIterator does not match formal parameter
found   : (Array<(adapt.cssvalid.Connection|null)>|null)
required: (Arguments<(adapt.cssvalid.Connection|null)>|Array<(adapt.cssvalid.Connection|null)>|Iterable<(adapt.cssvalid.Connection|null)>|Iterator<(adapt.cssvalid.Connection|null)>|string)
    for (const connection of this.connections) {
                             ^

src/adapt/cssvalid.js:1170: WARNING - actual parameter 1 of $jscomp.makeIterator does not match formal parameter
found   : (Array<string>|null)
required: (Arguments<string>|Array<string>|Iterable<string>|Iterator<string>|string)
        for (const name of this.propList) {
                           ^

src/adapt/cssvalid.js:1185: WARNING - actual parameter 1 of $jscomp.makeIterator does not match formal parameter
found   : (Array<string>|null)
required: (Arguments<string>|Array<string>|Iterable<string>|Iterator<string>|string)
    for (const name of this.propList) {
                       ^

src/adapt/cssvalid.js:1427: WARNING - actual parameter 1 of $jscomp.makeIterator does not match formal parameter
found   : (Array<string>|null)
required: (Arguments<string>|Array<string>|Iterable<string>|Iterator<string>|string)
    for (const name of this.propList) {
                       ^

src/adapt/cssvalid.js:2106: WARNING - actual parameter 1 of $jscomp.arrayFromIterable does not match formal parameter
found   : (Array<string>|null)
required: (Arguments<string>|Array<string>|Iterable<string>|string)
                        propList.push(...insetShorthand.propList);
                                         ^

src/adapt/cssvalid.js:2163: WARNING - actual parameter 1 of $jscomp.makeIterator does not match formal parameter
found   : (Array<string>|null)
required: (Arguments<string>|Array<string>|Iterable<string>|Iterator<string>|string)
    for (const prop of propList) {
                       ^

src/adapt/cssvalid.js:2167: WARNING - actual parameter 1 of $jscomp.makeIterator does not match formal parameter
found   : (Array<string>|null)
required: (Arguments<string>|Array<string>|Iterable<string>|Iterator<string>|string)
        for (const pname of list) {
                            ^

src/adapt/epub.js:121: WARNING - actual parameter 1 of $jscomp.makeIterator does not match formal parameter
found   : (Array<(null|string)>|null)
required: (Arguments<(null|string)>|Array<(null|string)>|Iterable<(null|string)>|Iterator<(null|string)>|string)
            for (const root of roots) {
                               ^

src/adapt/epub.js:564: WARNING - Misplaced function annotation. This JSDoc is not attached to a function node. Are you missing parentheses?
    class OPFDocumentURLTransformer {
    ^

src/adapt/epub.js:783: WARNING - actual parameter 1 of $jscomp.makeIterator does not match formal parameter
found   : (Array<(adapt.epub.OPFItem|null)>|null)
required: (Arguments<(adapt.epub.OPFItem|null)>|Array<(adapt.epub.OPFItem|null)>|Iterable<(adapt.epub.OPFItem|null)>|Iterator<(adapt.epub.OPFItem|null)>|string)
    for (const item of this.spine) {
                       ^

src/adapt/epub.js:1189: WARNING - Misplaced type annotation. Type annotations are not allowed here. Are you missing parentheses?
        /** @type {!adapt.epub.Position} */ {
                                            ^

src/adapt/epub.js:1804: WARNING - actual parameter 1 of $jscomp.makeIterator does not match formal parameter
found   : (NamedNodeMap<Attr>|null)
required: (Arguments<Attr>|Array<Attr>|Iterable<Attr>|Iterator<Attr>|string)
        for (const attr of attrs) {
                           ^

src/adapt/epub.js:1953: WARNING - actual parameter 1 of $jscomp.makeIterator does not match formal parameter
found   : (Array<{
  complete: boolean,
  instance: (adapt.ops.StyleInstance|null),
  item: (adapt.epub.OPFItem|null),
  layoutPositions: (Array<(adapt.vtree.LayoutPosition|null)>|null),
  pages: Array<(adapt.vtree.Page|null)>,
  xmldoc: (adapt.xmldoc.XMLDocHolder|null)
}>|null)
required: (Arguments<{
  complete: boolean,
  instance: (adapt.ops.StyleInstance|null),
  item: (adapt.epub.OPFItem|null),
  layoutPositions: (Array<(adapt.vtree.LayoutPosition|null)>|null),
  pages: Array<(adapt.vtree.Page|null)>,
  xmldoc: (adapt.xmldoc.XMLDocHolder|null)
}>|Array<{
  complete: boolean,
  instance: (adapt.ops.StyleInstance|null),
  item: (adapt.epub.OPFItem|null),
  layoutPositions: (Array<(adapt.vtree.LayoutPosition|null)>|null),
  pages: Array<(adapt.vtree.Page|null)>,
  xmldoc: (adapt.xmldoc.XMLDocHolder|null)
}>|Iterable<{
  complete: boolean,
  instance: (adapt.ops.StyleInstance|null),
  item: (adapt.epub.OPFItem|null),
  layoutPositions: (Array<(adapt.vtree.LayoutPosition|null)>|null),
  pages: Array<(adapt.vtree.Page|null)>,
  xmldoc: (adapt.xmldoc.XMLDocHolder|null)
}>|Iterator<{
  complete: boolean,
  instance: (adapt.ops.StyleInstance|null),
  item: (adapt.epub.OPFItem|null),
  layoutPositions: (Array<(adapt.vtree.LayoutPosition|null)>|null),
  pages: Array<(adapt.vtree.Page|null)>,
  xmldoc: (adapt.xmldoc.XMLDocHolder|null)
}>|string)
    for (const item of items) {
                       ^

src/adapt/epub.js:1969: WARNING - actual parameter 1 of $jscomp.makeIterator does not match formal parameter
found   : (Array<{
  complete: boolean,
  instance: (adapt.ops.StyleInstance|null),
  item: (adapt.epub.OPFItem|null),
  layoutPositions: (Array<(adapt.vtree.LayoutPosition|null)>|null),
  pages: Array<(adapt.vtree.Page|null)>,
  xmldoc: (adapt.xmldoc.XMLDocHolder|null)
}>|null)
required: (Arguments<{
  complete: boolean,
  instance: (adapt.ops.StyleInstance|null),
  item: (adapt.epub.OPFItem|null),
  layoutPositions: (Array<(adapt.vtree.LayoutPosition|null)>|null),
  pages: Array<(adapt.vtree.Page|null)>,
  xmldoc: (adapt.xmldoc.XMLDocHolder|null)
}>|Array<{
  complete: boolean,
  instance: (adapt.ops.StyleInstance|null),
  item: (adapt.epub.OPFItem|null),
  layoutPositions: (Array<(adapt.vtree.LayoutPosition|null)>|null),
  pages: Array<(adapt.vtree.Page|null)>,
  xmldoc: (adapt.xmldoc.XMLDocHolder|null)
}>|Iterable<{
  complete: boolean,
  instance: (adapt.ops.StyleInstance|null),
  item: (adapt.epub.OPFItem|null),
  layoutPositions: (Array<(adapt.vtree.LayoutPosition|null)>|null),
  pages: Array<(adapt.vtree.Page|null)>,
  xmldoc: (adapt.xmldoc.XMLDocHolder|null)
}>|Iterator<{
  complete: boolean,
  instance: (adapt.ops.StyleInstance|null),
  item: (adapt.epub.OPFItem|null),
  layoutPositions: (Array<(adapt.vtree.LayoutPosition|null)>|null),
  pages: Array<(adapt.vtree.Page|null)>,
  xmldoc: (adapt.xmldoc.XMLDocHolder|null)
}>|string)
    for (const item of items) {
                       ^

src/adapt/font.js:184: WARNING - actual parameter 1 of $jscomp.makeIterator does not match formal parameter
found   : (Array<(adapt.css.Val|null)>|null)
required: (Arguments<(adapt.css.Val|null)>|Array<(adapt.css.Val|null)>|Iterable<(adapt.css.Val|null)>|Iterator<(adapt.css.Val|null)>|string)
        for (const v of list) {
                        ^

src/adapt/font.js:355: WARNING - actual parameter 1 of $jscomp.makeIterator does not match formal parameter
found   : (Array<(adapt.font.Face|null)>|null)
required: (Arguments<(adapt.font.Face|null)>|Array<(adapt.font.Face|null)>|Iterable<(adapt.font.Face|null)>|Iterator<(adapt.font.Face|null)>|string)
    for (const srcFace of srcFaces) {
                          ^

src/adapt/geom.js:135: WARNING - actual parameter 1 of $jscomp.makeIterator does not match formal parameter
found   : (Array<(adapt.geom.Point|null)>|null)
required: (Arguments<(adapt.geom.Point|null)>|Array<(adapt.geom.Point|null)>|Iterable<(adapt.geom.Point|null)>|Iterator<(adapt.geom.Point|null)>|string)
    for (const p of this.points) {
                    ^

src/adapt/layout.js:59: WARNING - actual parameter 1 of $jscomp.makeIterator does not match formal parameter
found   : (Array<{
  bottom: number,
  height: number,
  left: number,
  right: number,
  top: number,
  width: number
}>|null)
required: (Arguments<{
  bottom: number,
  height: number,
  left: number,
  right: number,
  top: number,
  width: number
}>|Array<{
  bottom: number,
  height: number,
  left: number,
  right: number,
  top: number,
  width: number
}>|Iterable<{
  bottom: number,
  height: number,
  left: number,
  right: number,
  top: number,
  width: number
}>|Iterator<{
  bottom: number,
  height: number,
  left: number,
  right: number,
  top: number,
  width: number
}>|string)
    for (const box of boxes) {
                      ^

src/adapt/layout.js:126: WARNING - actual parameter 1 of $jscomp.makeIterator does not match formal parameter
found   : (Array<{
  bottom: number,
  height: number,
  left: number,
  right: number,
  top: number,
  width: number
}>|null)
required: (Arguments<{
  bottom: number,
  height: number,
  left: number,
  right: number,
  top: number,
  width: number
}>|Array<{
  bottom: number,
  height: number,
  left: number,
  right: number,
  top: number,
  width: number
}>|Iterable<{
  bottom: number,
  height: number,
  left: number,
  right: number,
  top: number,
  width: number
}>|Iterator<{
  bottom: number,
  height: number,
  left: number,
  right: number,
  top: number,
  width: number
}>|string)
        for (const box of boxes) {
                          ^

src/adapt/layout.js:959: WARNING - actual parameter 1 of $jscomp.makeIterator does not match formal parameter
found   : (Array<(adapt.geom.Band|null)>|null)
required: (Arguments<(adapt.geom.Band|null)>|Array<(adapt.geom.Band|null)>|Iterable<(adapt.geom.Band|null)>|Iterator<(adapt.geom.Band|null)>|string)
    for (const band of bands) {
                       ^

src/adapt/ops.js:245: WARNING - actual parameter 1 of $jscomp.makeIterator does not match formal parameter
found   : (Array<{condition: (adapt.expr.Val|null), properties: (adapt.csscasc.ElementStyle|null)}>|null)
required: (Arguments<{condition: (adapt.expr.Val|null), properties: (adapt.csscasc.ElementStyle|null)}>|Array<{condition: (adapt.expr.Val|null), properties: (adapt.csscasc.ElementStyle|null)}>|Iterable<{condition: (adapt.expr.Val|null), properties: (adapt.csscasc.ElementStyle|null)}>|Iterator<{condition: (adapt.expr.Val|null), properties: (adapt.csscasc.ElementStyle|null)}>|string)
    for (const fontFace of self.style.fontFaces) {
                           ^

src/adapt/ops.js:386: WARNING - actual parameter 1 of $jscomp.makeIterator does not match formal parameter
found   : (Array<(adapt.vtree.FlowChunkPosition|null)>|null)
required: (Arguments<(adapt.vtree.FlowChunkPosition|null)>|Array<(adapt.vtree.FlowChunkPosition|null)>|Iterable<(adapt.vtree.FlowChunkPosition|null)>|Iterator<(adapt.vtree.FlowChunkPosition|null)>|string)
        for (const p of flowPosition.positions) {
                        ^

src/adapt/ops.js:403: WARNING - Misplaced type annotation. Type annotations are not allowed here. Are you missing parentheses?
                /** @type {boolean} */ new adapt.expr.Named(this.style.pageScope, `${side}-page`).evaluate(this)
                                       ^

src/adapt/ops.js:403: WARNING - inconsistent return type
found   : (boolean|number|string|undefined)
required: boolean
                /** @type {boolean} */ new adapt.expr.Named(this.style.pageScope, `${side}-page`).evaluate(this)
                                       ^

src/adapt/pm.js:1406: WARNING - actual parameter 1 of $jscomp.makeIterator does not match formal parameter
found   : (Array<(adapt.pm.PageBox|null)>|null)
required: (Arguments<(adapt.pm.PageBox|null)>|Array<(adapt.pm.PageBox|null)>|Iterable<(adapt.pm.PageBox|null)>|Iterator<(adapt.pm.PageBox|null)>|string)
    for (const child of this.pageBox.children) {
                        ^

src/adapt/sha1.js:118: WARNING - actual parameter 1 of $jscomp.makeIterator does not match formal parameter
found   : (Array<number>|null)
required: (Arguments<number>|Array<number>|Iterable<number>|Iterator<number>|string)
    for (const n of h) {
                    ^

src/adapt/task.js:369: WARNING - actual parameter 1 of $jscomp.makeIterator does not match formal parameter
found   : (Array<function (): undefined>|null)
required: (Arguments<function (): undefined>|Array<function (): undefined>|Iterable<function (): undefined>|Iterator<function (): undefined>|string)
            for (const callback of task.callbacks) {
                                   ^

src/adapt/xmldoc.js:528: WARNING - actual parameter 1 of $jscomp.makeIterator does not match formal parameter
found   : (Array<Node>|null)
required: (Arguments<Node>|Array<Node>|Iterable<Node>|Iterator<Node>|string)
    for (const n of this.nodes) {
                    ^

src/vivliostyle/counters.js:102: WARNING - Misplaced function annotation. This JSDoc is not attached to a function node. Are you missing parentheses?
    class CounterListener {
    ^

src/vivliostyle/counters.js:132: WARNING - Misplaced function annotation. This JSDoc is not attached to a function node. Are you missing parentheses?
    class CounterResolver {
    ^

src/vivliostyle/counters.js:656: WARNING - Misplaced function annotation. This JSDoc is not attached to a function node. Are you missing parentheses?
    class LayoutConstraint {
    ^

src/vivliostyle/layoututil.js:490: WARNING - actual parameter 1 of vivliostyle.layoututil.LayoutMode.accept does not match formal parameter
found   : (adapt.vtree.NodeContext|null)
required: adapt.vtree.NodeContext
            let accepted = mode.accept(positionAfter, column);
                                       ^

src/vivliostyle/layoututil.js:491: WARNING - actual parameter 1 of vivliostyle.layoututil.LayoutMode.postLayout does not match formal parameter
found   : (adapt.vtree.NodeContext|null)
required: adapt.vtree.NodeContext
            accepted = mode.postLayout(positionAfter, this.initialPosition, column, accepted);
                                       ^

src/vivliostyle/table.js:1445: WARNING - actual parameter 1 of adapt.vtree.ClientLayout.getElementClientRect does not match formal parameter
found   : (Node|null)
required: (Element|null)
            const tableBBox = column.clientLayout.getElementClientRect(tableElement);
                                                                       ^

src/vivliostyle/table.js:1455: WARNING - actual parameter 2 of vivliostyle.table.TableLayoutProcessor.prototype.normalizeColGroups does not match formal parameter
found   : (Node|null)
required: Element
            this.normalizeColGroups(formattingContext, tableElement, column);
                                                       ^

src/vivliostyle/viewerapp.js:106: WARNING - Property DOM_KEY_LOCATION_NUMPAD never defined on KeyboardEvent
        || keyIdentifier === "U+00BB" || (keyIdentifier === "U+004B" && location === KeyboardEvent.DOM_KEY_LOCATION_NUMPAD) /* workaround for Chrome for Windows */) {
                                                                                                   ^

src/vivliostyle/viewerapp.js:111: WARNING - Property DOM_KEY_LOCATION_NUMPAD never defined on KeyboardEvent
        || keyIdentifier === "U+00BD" || (keyIdentifier === "U+004D" && location === KeyboardEvent.DOM_KEY_LOCATION_NUMPAD) /* workaround for Chrome for Windows */) {
                                                                                                   ^

0 error(s), 46 warning(s), 97.46505779036049% typed
```

</details>

File: src/adapt/css.js [hunk 3]
```diff
@@ -165,84 +165,69 @@ goog.inherits(adapt.css.FilterVisitor, adapt.css.Visitor);
...
-        } else if (before !== after) {
-            arr = new Array(values.length);
-            for (var k = 0; k < i; k++) {
-                arr[k] = values[k];
-            }
-            arr[i] = after;
-        }
-    }
-    return arr || values;
+ /** @type {Array.<adapt.css.Val>} */ let arr = null;
```
[the remark this answers is not in the document]
Comment (MurakamiShinyu) [reply src/adapt/css.js]: I don't know why `lebab --transform for-each` causes this discrepant indent, and also I think this "for-each" transformation result code is not better than before.
Should we revert [this "for-each" transformation](https://github.com/MurakamiShinyu/vivliostyle.js/commit/238808ca99fbf688e09fc8e56431129d93be78ab#diff-84d811d66ef3c951f0302b4075315866L167)?

[the remark this answers is not in the document]
Comment (spring-raining) [reply src/adapt/css.js]: In terms of code readability, these conversions are not very big difference. We might have better to preserve previous codes.

**Pull request merged**